File: LeetCode/Array/Golang/Leetcode798-得分最高的最小轮调.go
package sbgo

func bestRotation(nums []int) int {
	n := len(nums)
	b := make([]int, n+1)
	for i := 0; i < n; i++ {
		l, r := i-nums[i]+1, i
		if l >= 0 {
			b[l]++
			b[r+1]--
		} else {
			b[0], b[r+1] = b[0]+1, b[r+1]-1
			b[l+n], b[n] = b[l+n]+1, b[n]-1
		}
	}
	res, k := int(1e6), -1
	for i, sum := 0, 0; i < n; i++ {
		sum += b[i]
		if res > sum {
			res = sum
			k = i
		}
	}
	return k
}

File: LeetCode/Array/JS/Leetcode798-得分最高的最小轮调.js
var bestRotation = function (nums) {
    const n = nums.length;
    const b = new Array(n + 1).fill(0);
    for (let i = 0; i < n; i++) {
        const l = i - nums[i] + 1, r = i;
        if (l >= 0) {
            b[l]++, b[r + 1]--;
        } else {
            b[0]++, b[r + 1]--;
            b[l + n]++, b[n]--;
        }
    }
    let res = Math.floor(1e6 + 10), k = 0;
    for (let i = 0, sum = 0; i < n; i++) {
        sum += b[i];
        if (res > sum) {
            res = sum;
            k = i;
        }
    }
    return k;
};